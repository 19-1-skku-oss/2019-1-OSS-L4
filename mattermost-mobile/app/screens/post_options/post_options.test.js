// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {Alert} from 'react-native';
import {shallow} from 'enzyme';

import Preferences from 'mattermost-redux/constants/preferences';

import PostOptions from './post_options';

jest.mock('react-intl');

jest.mock('Alert', () => {
    return {
        alert: jest.fn(),
    };
});

describe('PostOptions', () => {
    const navigator = {
        showModal: jest.fn(),
        dismissModal: jest.fn(),
        push: jest.fn(),
    };

    const actions = {
        addReaction: jest.fn(),
        deletePost: jest.fn(),
        flagPost: jest.fn(),
        pinPost: jest.fn(),
        removePost: jest.fn(),
        unflagPost: jest.fn(),
        unpinPost: jest.fn(),
    };

    const post = {
        root_id: 'root_id',
        id: 'post_id',
        message: 'message',
        is_pinned: false,
        channel_id: 'channel_id',
    };

    const baseProps = {
        actions,
        canAddReaction: true,
        canReply: true,
        canDelete: true,
        canPin: true,
        canEdit: true,
        canEditUntil: -1,
        channelIsReadOnly: false,
        currentTeamUrl: 'http://localhost:8065/team-name',
        deviceHeight: 600,
        hasBeenDeleted: false,
        isFlagged: false,
        isMyPost: true,
        isSystemMessage: false,
        managedConfig: {},
        navigator,
        post,
        showAddReaction: true,
        theme: Preferences.THEMES.default,
    };

    function getWrapper(props = {}) {
        return shallow(
            <PostOptions
                {...baseProps}
                {...props}
            />,
            {context: {intl: {formatMessage: ({defaultMessage}) => defaultMessage}}}
        );
    }

    test('should match snapshot, showing all possible options', () => {
        const wrapper = getWrapper();

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('should match snapshot, showing Delete option only for system message to user who has permission to delete', () => {
        const wrapper = getWrapper({isSystemMessage: true});

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('should match snapshot, no option for system message to user who doesn\'t have the permission to delete', () => {
        const wrapper = getWrapper({isSystemMessage: true, canDelete: false});

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('should load thread', () => {
        const wrapper = getWrapper();
        const instance = wrapper.instance();

        instance.closeWithAnimation = jest.fn();

        wrapper.findWhere((node) => node.key() === 'reply').simulate('press');
        expect(instance.closeWithAnimation).toBeCalled();
    });

    test('should not show reply option', () => {
        const wrapper = getWrapper({canReply: false});

        expect(wrapper.findWhere((node) => node.key() === 'reply')).toMatchObject({});
    });

    test('should remove post after delete', () => {
        const wrapper = getWrapper();

        wrapper.findWhere((node) => node.key() === 'delete').simulate('press');
        expect(Alert.alert).toBeCalled();

        // Trigger on press of Delete in the Alert
        Alert.alert.mock.calls[0][2][1].onPress();

        expect(actions.deletePost).toBeCalled();
        expect(actions.removePost).toBeCalled();
    });
});
