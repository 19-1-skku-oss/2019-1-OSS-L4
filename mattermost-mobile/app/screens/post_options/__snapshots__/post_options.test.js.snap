// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostOptions should match snapshot, no option for system message to user who doesn't have the permission to delete 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ForwardRef(forwardConnectRef)
    allowStayMiddle={false}
    initialPosition={250}
    marginFromTop={350}
    onRequestClose={[Function]}
  >
    <PostOption
      icon="edit"
      onPress={[Function]}
      text="Edit"
    />
    <PostOption
      icon="reply"
      onPress={[Function]}
      text="Reply"
    />
    <PostOption
      icon="pin"
      onPress={[Function]}
      text="Pin to Channel"
    />
    <PostOption
      icon="emoji"
      onPress={[Function]}
      text="Add Reaction"
    />
  </ForwardRef(forwardConnectRef)>
</View>
`;

exports[`PostOptions should match snapshot, showing Delete option only for system message to user who has permission to delete 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ForwardRef(forwardConnectRef)
    allowStayMiddle={false}
    initialPosition={300}
    marginFromTop={300}
    onRequestClose={[Function]}
  >
    <PostOption
      icon="edit"
      onPress={[Function]}
      text="Edit"
    />
    <PostOption
      icon="reply"
      onPress={[Function]}
      text="Reply"
    />
    <PostOption
      icon="pin"
      onPress={[Function]}
      text="Pin to Channel"
    />
    <PostOption
      icon="emoji"
      onPress={[Function]}
      text="Add Reaction"
    />
    <PostOption
      destructive={true}
      icon="trash"
      onPress={[Function]}
      text="Delete"
    />
  </ForwardRef(forwardConnectRef)>
</View>
`;

exports[`PostOptions should match snapshot, showing all possible options 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ForwardRef(forwardConnectRef)
    allowStayMiddle={false}
    initialPosition={300}
    marginFromTop={300}
    onRequestClose={[Function]}
  >
    <PostOption
      icon="edit"
      onPress={[Function]}
      text="Edit"
    />
    <PostOption
      icon="reply"
      onPress={[Function]}
      text="Reply"
    />
    <PostOption
      icon="pin"
      onPress={[Function]}
      text="Pin to Channel"
    />
    <PostOption
      icon="emoji"
      onPress={[Function]}
      text="Add Reaction"
    />
    <PostOption
      destructive={true}
      icon="trash"
      onPress={[Function]}
      text="Delete"
    />
  </ForwardRef(forwardConnectRef)>
</View>
`;
